File: api/Data/Models/LeagueSummary.cs
using System.Text.Json.Serialization;
using FootballGm.Api.Data.Enums;
using FootballGm.Api.Serialization;
using Entities = FootballGm.Api.Data.Entity.Contrived;

namespace FootballGm.Api.Data.Models;

public sealed class LeagueSummary
{
    public required int LeagueId { get; init; }
    public required string Name { get; init; }
    public required string JoinCode { get; init; }

    [JsonConverter(typeof(CamelCaseEnumConverter<LeagueMemberRole>))]
    public required LeagueMemberRole Role { get; init; }

    [JsonConverter(typeof(CamelCaseEnumConverter<ScoringKind>))]
    public required ScoringKind Scoring { get; init; }

    public static LeagueSummary From(League league, LeagueMemberRole role) => new()
    {
        LeagueId = league.LeagueId,
        Name = league.Name,
        JoinCode = league.JoinCode,
        Role = role,
        Scoring = Rule.UsesDefaultScoringWeights(league.Rules)
            ? ScoringKind.Standard
            : ScoringKind.Custom
    };

    public static LeagueSummary From(Entities.League entity, LeagueMemberRole role) =>
        From(League.FromEntity(entity), role);
}

File: app/lib/features/leagues/models/league_summary.dart
/// Slim list row for a League the User belongs to.
class LeagueSummary {
  const LeagueSummary({
    required this.leagueId,
    required this.name,
    required this.joinCode,
    required this.role,
    required this.scoring,
  });

  final int leagueId;
  final String name;
  final String joinCode;
  final LeagueRole role;
  final ScoringKind scoring;

  factory LeagueSummary.fromJson(Map<String, dynamic> json) {
    return LeagueSummary(
      leagueId: json['leagueId'] as int,
      name: json['name'] as String,
      joinCode: json['joinCode'] as String,
      role: LeagueRole.parse(json['role'] as String),
      scoring: ScoringKind.parse(json['scoring'] as String),
    );
  }
}

enum LeagueRole {
  member,
  admin,
  commissioner;

  String get label => switch (this) {
        member => 'Member',
        admin => 'Admin',
        commissioner => 'Commissioner',
      };

  static LeagueRole parse(String value) => switch (value) {
        'member' => member,
        'admin' => admin,
        'commissioner' => commissioner,
        _ => throw FormatException('Unknown role: $value'),
      };
}

enum ScoringKind {
  standard,
  custom;

  String get label => switch (this) {
        standard => 'Standard scoring',
        custom => 'Custom scoring',
      };

  static ScoringKind parse(String value) => switch (value) {
        'standard' => standard,
        'custom' => custom,
        _ => throw FormatException('Unknown scoring: $value'),
      };
}
